### ISSUE TITLE
Campaign Runs

### ISSUE THREAD

**Issue by DFurnes** [DoSomething/phoenix-next#18]
We currently use "campaign runs" to close and re-launch campaigns year after year. This works okay, but complicates our data model (nested relationships and more IDs to send around), and also seems to be a common source of confusion for the rest of the organization.

### History
I keep forgetting why we ever made runs, so hopefully this helps provide context:

1. We first launched campaigns without runs.
2. A few months later, we realized we needed to "close" Comeback Clothes, so we introduced "campaign runs" as a way to hold all the closed state fields for a campaign (press, featured RBs...)
3. About a year later, we realized we needed a way to re-run campaigns a second time... so the ability to "archive" signups & reportbacks onto a run was added. This would transfer activity data off the campaign and onto the run as an archive.
4. Then we realized our folly once again when we couldn't close a campaign a second time. This is when we added the ability to attach multiple runs to a single campaign, and the signups & reportbacks were updated to always live on the runs.

### Discussion
Two suggestions that came up in our meeting the other day:

__Can we just remove runs altogether and always create a new campaign the following year?__ I think part of why we used runs as a solution in Drupal is because duplicating campaigns was hard/impossible. We don't have that problem in Contentful. So instead of having a "Teens For Jeans" campaign with "T4J 2016" and "T4J 2017" runs, we would just have two completely separate campaigns.

__For "closed state" fields, would introducing a "Wrap Report" nested content type make more sense?__ This would help us avoid hitting the 50 field limit in Contentful, and seems like what runs were initially intended to be.

---

__tl;dr:__ Do we still need these?

**Comment by angaither** [comment]
Two other issues in a _life without runs_
1. keeping the URL the same year over year (certainly other ways around this)
2. holding historical data for the same action together as a group (again, other ways around this)
> over 4 years we had XX jeans collected, and last year, our members collected YY

**Comment by deadlybutter** [comment]
> holding historical data for the same action together as a group (again, other ways around this)

Is this a Quasar request, or do we plan on requiring this functionality in Phoenix?

**Comment by sheyd** [comment]
Does Contentful have the concept of changeable vanity URI's to address this?

> keeping the URL the same year over year (certainly other ways around this)

**Comment by weerd** [comment]
@sheyd I don't think we'd need to if we take a particular approach w/ how we create content types in Contentful. I've already played around with this. Contentful doesn't handle anything w/ URI's for content, because it's just an API to query and retrieve content. We deal w/ the URI's on our own.

**Comment by deadlybutter** [comment]
@sheyd Are you able to offer any insight on the data concerns? What kind of queries do you currently/anticipate running that involve year over year campaigns?

**Comment by angaither** [comment]
@deadlybutter I could see it as both a data request and member facing info
"Here's what we all accomplished together last time" kind of deal

**Comment by deadlybutter** [comment]
gotcha!

**Comment by sheyd** [comment]
@deadlybutter The biggest concern on the data front, in so far as the data model is still being worked out is mostly with whether we can extract unique campaigns with all associated content that has enough metadata context to figure out what it maps to. To use TFJ example: if TFJ2016 and then TFJ2017 keep same API campaign URI, on the data side, that would corrupt things, so we'd have to have some unique identifier or identifiers (whatever those might be), to identify the campaign that happened in 2016 and 2017 separately. And ideally this would be automatically populated by content creation process, i.e. content creator or admin wouldn't have to think about updating a field manually every time a campaign was potentially re-used.

**Comment by weerd** [comment]
> if TFJ2016 and then TFJ2017 keep same API campaign URI, on the data side, that would corrupt things, so we'd have to have some unique identifier or identifiers (whatever those might be), to identify the campaign that happened in 2016 and 2017 separately.

@sena would keeping the same url `/campaigns/teens-for-jeans` and having the unique identifier be some sort of timestamp (year/full date, etc) still work? So when you want data on an instance of TFJ, you, the url stays the same and the query includes the year identifier to know you want data from TFJ2016 and not 2017?

**Comment by deadlybutter** [comment]
@sheyd Assuming we hade a way to "link" all the campaigns together as a run, could you use creation date as the unique identifier?

**Comment by deadlybutter** [comment]
AH diego beat me to it

**Comment by sheyd** [comment]
@weerd @deadlybutter Yes, with further questions:

1. Would there be a way to query *all* campaign URI's, e.g. https://api.contentful.com/ds/all-campaigns, so we wouldn't have to keep a manual list of all the campaigns we had ever created stored somewhere (unless it's being mapped to another layer of our content platform potentially).
2. How unique would timestamp be? If just year, we run the risk of, let's say a campaign we want to run multiple times in a year with same URI being over-written and data lost.
3. What other mechanism does Contenful have to uniquely identify campaign (at least the way we're using them)?

**Comment by deadlybutter** [comment]
@sheyd 

1. Yep! https://www.contentful.com/developers/docs/references/content-management-api/#/reference/entries/entries-collection/get-all-entries-of-a-space

2. In my head I was assuming it was a full ISO timestamp (Contentful automatically makes this in the metadata), so safe to assume it's unique enough for our purposes

3. We can technically add any unique field we want, though right now the only two I _know of in the current prototype model_ are `id` & `slug`

**Comment by weerd** [comment]
@sheyd Yeah, like @deadlybutter mentions, we can query to get all entries, specifically all Campaign entries so that wouldn't be an issue. 

So the way I'm thinking about it, wrap-up information for a timeboxed instance of a campaign run would be it's own content type (@DFurnes mentioned an idea like this yesterday and I'm trying to roll with it cause it sounds rather doable). Each of those referenced content type "wrap-up report runs" would contain a timestamp (and it's own uuid (mongo-style id)) that would uniquely indicate, something to the effect of "this is the full wrap-up data for the associated campaign for this _insert-timestamp-here_ run".

**Comment by deadlybutter** [comment]
> Each of those referenced content type

Sorry, nitpicky, but if we're duplicating campaigns we would have have a singular reference to a wrap report, right?

**Comment by weerd** [comment]
@deadlybutter I'm not talking about duplicating campaigns. The campaign stays as a single content item and thus eliminates the issue w/ the URI changing.

![wrap-up content type idea](https://cloud.githubusercontent.com/assets/105849/21729365/c7d8a1ac-d418-11e6-9adc-61b10893f79a.png)

So in the example visual above, there is a **Campaign Run Wrap-Up** content type and we can add those per each run, regardless if a campaign happens more than once per year (I'm using just "years" in the example visual, but consider it a full on timestamp). So it's a modification of the "runs" idea, but not duplicating all the content etc, and needing to deal w/ URI issues. The wrap-up content type _just_ handles the information we would care about for depicting the end result of an instance run of a campaign.

**Comment by deadlybutter** [comment]
(Love the visuals, need to bring my iPad more)

So If I run TFJ a second time, I'm going to overwrite the content from last year & just attach a new wrap report when it's over?

**Comment by weerd** [comment]
@deadlybutter yep yep. _IF_ the content even needs to be overwritten, maybe it doesn't... (being realistic, probs most of it doesn't) AND if we ever really _need_ to refer to some old content on a campaign (what's the reality of us ever needing to do that), we have revisions/versions in Contentful to look that up if truly necessary 😉

**Comment by deadlybutter** [comment]
@weerd agreed, makes sense

**Comment by weerd** [comment]
It feels like it covers the bases we want, and also simplifies things and only saves the data we _really_ care about per each campaign run.

**Comment by sheyd** [comment]
@weerd One small nitpick note about the *data we really care about*. You never know when you'll want data for something or it might come in handy, so whatever we can do to have full fidelity of data whether things are re-used/changed, etc. is appreciated.

**Comment by weerd** [comment]
@sheyd yeah, I think that would apply to any solution within Contentful that we come up with. So we'll need to figure something out regardless, but I think we can use a JSON Object field in the _Campaign Run Wrap-up_ content type as a catch-all for very custom data, unique to each run :)

**Comment by deadlybutter** [comment]
something i just thought of, since Quasar is SQL (@sheyd ?), our catch-all JSON might not be stored there (Unless Sena plans on creating columns for every possible custom setting, which seems crazy?). I think we should assume anything we want to later ask questions about in data should actually be a part of the content model in some fashion.

**Comment by DFurnes** [comment]
> So in the example visual above, there is a Campaign Run Wrap-Up content type and we can add those per each run, regardless if a campaign happens more than once per year (I'm using just "years" in the example visual, but consider it a full on timestamp).

@weerd This makes decent sense to me (and +1 to the visualization being great).

One concern: how would signups/reportbacks for different "years" of a campaign be differentiated in Rogue? Would we have to query those by timestamp of the most recent "wrap-up" report?

**Comment by weerd** [comment]
@DFurnes yeah, not 100% certain on this. It could be the timestamp window of the `started_at` and `ended_at` for the run wrap-up. It could also just be the run wrap-up `id`, although if this is going to be easy to use, I don't think a run wrap-up should have to be created _before_ the campaign re-opens, but only when it's about to close (I feel this would be easier, I dunno). And if that's the case then the id for the run wrap-up that RBs should be assigned wouldn't exist if it's created at the end of the campaign run.

There might be some other alternatives not coming to mind at the moment...

**Comment by deadlybutter** [comment]
@weerd Is there a way in Contentful to make like a "button" which triggers a webhook? Every time we re-open a campaign, an admin clicks the "button" which tells Phoenix to generate a new run. 

On second thought, is it possible to make this a custom JS/CSS plugin? Make a custom field called "run_id" but style it as a button. Every time it's clicked the JS will generate a UUID & update the field value.

**Comment by weerd** [comment]
@deadlybutter there are webhooks in Contentful, but not sure if they're more exclusively for your connecting app to hit them... 

Sounds like something that could work, but I'd prefer to work through ideas first that involve a simpler fix if possible (like referring to timestamps from runs for the assoc RBs) so it's less confusion. If it turns out none of those are feasible, then something a bit more complicated will have to suffice.

I was thinking of something similar, like when a campaign is re-opened (or created for the first time) a webhook creates an inactive run wrap-up like you mention and thus we'd have an ID for it to assoc w/ RBs...

**Comment by DFurnes** [comment]
Hmm, this is why I'm partial to _only_ having campaigns rather than a nested campaign/run relationship. It seems to end up with more complex editorial rules or backend logic otherwise.

How about if we had Campaigns & Wrap Reports, but only a one-to-one relationship. When we want to re-run a campaign later on, the campaign lead can duplicate the campaign & swap it's URL with the old one. For data purposes, we could always have a queryable "bucket" (e.g. `teens-for-jeans` on all Teens For Jeans campaigns).

**Comment by deadlybutter** [comment]
So when I (campaign lead) create a campaign for the first time, I

1) Set a public facing title
2) Set an immutable internal title (or "bucket", also fine for me)

When I go to re-run a campaign,
1) Duplicate campaign
2) Keep or Alter the public title, its wouldn't matter
3) Keep the same internal bucket name

**Comment by DFurnes** [comment]
I think this was the loose consensus from our meeting today: every campaign has a __public title__ (e.g. "Teens for Jeans"), an __internal title__ (e.g. "Teens for Jeans 2017"), and a __bucket__ based on the cause space or call-to-action (e.g. "homelessness-collect-jeans").

The bucket would be a field stored on the campaign, and would allow the data team or internal services to query for all campaigns that are assigned to a particular bucket (for example, "find me all campaign IDs for `homelessness-collect-jeans` campaigns").

**Comment by deadlybutter** [comment]
Something I got confused on was the reason behind the "Teens for jeans 2017". Every campaign would have a timestamp already, why write this twice?

**Comment by angaither** [comment]
> all campaigns that are assigned to a particular bucket (for example, "find me all campaign IDs for homelessness-collect-jeans campaigns").

This brings up an interesting thought in my brain, do we run multiple cause space noun/verb combo campaigns?

**Comment by DFurnes** [comment]
> Something I got confused on was the reason behind the "Teens for jeans 2017". Every campaign would have a timestamp already, why write this twice?

I think it's nice for the Contentful admin view. A list of "Teens For Jeans", "Teens For Jeans", "Teens For Jeans", "Teens For Jeans" only ordered by timestamp may be visually confusing.

> This brings up an interesting thought in my brain, do we run multiple cause space noun/verb combo campaigns?

Hmm, we probably do – clothing donation campaign or card-making campaigns have existed for multiple sponsors right? And I guess we'd want those to be completely separate buckets if so...

**Comment by deadlybutter** [comment]
add a unique word at the end of it?

`homelessness-collect-jeans-[something]`
